demos/object_detection_demo_ssd_async/main.cpp

```diff
@@ -354,58 +403,194 @@ int main(int argc, char *argv[]) {
+                auto startTime = std::chrono::steady_clock::now();
+                
+                cv::Mat frame;
+                if (!cap.read(frame)) {
+                    if (frame.empty()) {
+                        if (FLAGS_loop_input) {
+                            cap.open((FLAGS_i == "cam") ? 0 : FLAGS_i.c_str());
+                        } else cap.release();
+                        continue;
+                    } else {
+                        throw std::logic_error("Failed to get frame from cv::VideoCapture");
+                    }
+                }
 
-            if (isModeChanged) {
-                isModeChanged = false;
-            }
+                InferRequest::Ptr request;
+                {
+                    std::lock_guard<std::mutex> lock(mutex);
 
-            // Final point:
-            // in the truly Async mode we swap the NEXT and CURRENT requests for the next iteration
-            curr_frame = next_frame;
-            next_frame = cv::Mat();
-            if (isAsyncMode) {
-                async_infer_request_curr.swap(async_infer_request_next);
-            }
+                    request = std::move(emptyRequests.front());
+                    emptyRequests.pop_front();
+                }
+                frameToBlob(frame, request, imageInputName);
+                
+                bool frameMode = isUserSpecifiedMode;
+                request->SetCompletionCallback([&mutex,
+                                                &completedRequestResults,
+                                                nextFrameId,
+                                                frame,
+                                                request,
+                                                &outputName,
+                                                startTime,
+                                                frameMode,
+                                                &isUserSpecifiedMode,
+                                                &emptyRequests,
+                                                &callbackException,
+                                                &condVar] {
+                    {
+                        std::lock_guard<std::mutex> callbackLock(mutex);
+                    
+                        try {
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame,
+                                    std::unique_ptr<LockedMemory<const void>>(new
+                                        LockedMemory<const void>(as<MemoryBlob>(request->GetBlob(outputName))->rmap())),
+                                    startTime,
+                                    frameMode == isUserSpecifiedMode
+                                }));
+                            
+                            if (isUserSpecifiedMode == frameMode) {
```

Comment:
I don't quite understand why this condition is here. Why wouldn't we want to return `request` to `emptyRequests` if the execution mode changed?

Reply:
It's because the `emptyRequests` container is only for requests from current mode. I've changed the name to make it clearer.

Reply:
> It's because the emptyRequests container is only for requests from current mode.

Okay, but why, though? When you switch modes, you wait for all pending requests to complete and then reinitialize `emptyRequests`, so does it even matter which requests it contains at that point?